### PR TITLE
Fix evaluated and weighted_objective

### PR DESCRIPTION
**Pull request by cangermueller**
Hi all,

first: Keras is awesome :-) !
I am new here, and suggest the following PR to make Keras even better:

1) Fixes weight standardization in `evaluate()` of Graph model
Before, the unnormalized version of output vectors (y) was used to standardize weights, which resulted in a shape mismatch between y and the weight vector when standardize_y changed the shape of y. Now, y is first standardized and used afterwards to standardize weights as also done in other methods like `fit()`.

2) Fixes division by zero in `weighted_objective()`
Before, `weighted_objective()` returned nan due to division by zero if the weight vector contained only zero weights. Now, it returns zero if there are no observations. I found this bug when I trained a multi-task model with sparse output vectors. In some batches, there were no observations for some tasks, which resulted in nan in the objective.

Cheers,
Christof

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -73,10 +73,10 @@ def weighted(y_true, y_pred, weights, mask=None):
         weighted = filtered_weights * obj_output
         if mask is None:
             # Instead of calling mean() here, we divide by the sum of filtered_weights.
-            return weighted.sum() / filtered_weights.sum()
+            return weighted.sum() / (filtered_weights.sum() + 1e-20)
```
Comment (fchollet) [inline keras/models.py]: If we were to use a epsilon here... it should be abstracted as a globally configured epsilon. Also, I think this value is too low and would cause issues with GPUs (float32).

But anyway: we shouldn't. If the sum of the weights is zero, then the problem is ill-defined and a loss cannot be computed. We should throw an exception in that case.

**Comment by cangermueller** [comment]
I agree that the **average weighted loss** 1/W \sum_i=1^N f(y_i, y’_i) is undefined mathematically. 

However, I suggest to add a globally defined epsilon that does not result in an underflow on GPUs.

1) It’s better to return 0 instead of throwing and exception due to practical reasons. 
Lets say y = [0 1 1 x 0 x], x means unobserved (weight=0), and the batch size is 3. Then the function won’t throw an exception most of the time, but in the unlucky shuffle split y’=[0 1| 0 1 | x x] it will. At the end, the user does not want the training to be interrupted just because it happens that there are no observations in one batch. It might also be hard to reproduce if the user does not set a seed at the beginning. Finally, the loss over the whole vector without batch calculations is defined as well.

2) The **weighted loss** \sum_i=1^N f(y_i, y’_i) is 0, if N = 0. If there are no observations, then the loss and the gradient should be 0 and there is no reasons to update model parameters.

- Christof

File: keras/models.py [hunk 2]
```diff
@@ -73,10 +73,10 @@ def weighted(y_true, y_pred, weights, mask=None):
...
         else:
             filtered_mask = mask[weights.nonzero()[:-1]]
-            return weighted.sum() / (filtered_mask * filtered_weights).sum()
+            return weighted.sum() / ((filtered_mask * filtered_weights).sum() + 1e-20)
```
Comment (EderSantana) [inline keras/models.py]: please, see a few lines above, all the `sample_weight` values that are equal zero are filtered out before hand. Unless somebody changed that later.

Also, this is overlapping with what we are building for the objective decorator. We will give more control to the user about how he wants to use his sample weights. We wont be forcing anything. See #802

**Comment by cangermueller** [comment]
Yes, zero weights are filtered out. The problem occurs when no non-nonzero values remain, i.e. when `weights.nonzero()` only contains `False` such that the sum is 0. Or did did I misunderstand you?

**Comment by EderSantana** [comment]
`sample_weights` are supposed to be all positive values and never sum to zero. Think of them as a (unnormalized) probability distribution over the samples. They define how much each sample contributes to the cost function. In which case you found `sample_weights` summing to zero? Do you have samples that are supposed to maximize instead of minimize the loss function?

**Comment by cangermueller** [comment]
`sample_weights` for all samples, e.g. used to call `fit()` or `evaluate()`, might have positive values, but the `weights` vector used to call `weighted(y_true, y_pred, weights, mask=None)` can have only zero values anyhow due to batch-wise computations as described above. `sample_weights` is randomly split into chunks, e.g. in `_fit()` or `_test_loop()`.

**Comment by EderSantana** [comment]
I see, but if all weights are zero, isn't it better to avoid the gradiente computation at all? for example, if weights.nonzero() is None, then `return cost = 0`. No need to regularize a cost that doesn't exist.

**Comment by cangermueller** [comment]
Yes, `return 0 if filtered_weights.sum() == 0` what be almost the same as `return weighted.sum() / (filtered_weights.sum() + 1e-20)`, since $$0 \approx 0 / (1 + 1e-20)$$ ;-)

**Comment by EderSantana** [comment]
Makes sense, I understand now. I'll also make this case is covered by the new cost function API

**Comment by cangermueller** [comment]
Should I `return 0 if filtered_weights.sum() == 0`, or will you merge the PR as it is?

**Comment by fchollet** [comment]
> Should I return 0 if filtered_weights.sum() == 0, or will you merge the PR as it is?

It sounds much cleaner. Introducing the epsilon raises a wide range of issues.

**Comment by cangermueller** [comment]
I replaced epsilon by an ifelse condition that checks if the sum of weights is zero.

**Comment by fchollet** [comment]
This has been fixed in the latest release.
